rfc: database storage engine (#6897)
tendermint/tendermint: docs/rfc/rfc-001-storage-engine.rst

===========================================
RFC 001: Storage Engines and Database Layer
===========================================

Changelog
---------

- 2021-04-19: Initial Draft (gist)
- 2021-09-02: Migrated to RFC folder, with some updates  

Abstract
--------

The aspect of Tendermint that's responsible for persistence and storage (often
"the database" internally) represents a bottle neck in the architecture of the
platform, that the 0.36 release presents a good opportunity to correct. The
current storage engine layer provides a great deal of flexibility that is
difficult for users to leverage or benefit from, while also making it harder
for Tendermint Core developers to deliver improvements on storage engine. This
RFC discusses the possible improvements to this layer of the system.

Background
----------

Tendermint has a very thin common wrapper that makes Tendermint itself
(largely) agnostic to the data storage layer (within the realm of the popular
key-value/embedded databases.) This flexibility is not particularly useful:

> the benefits of a specific database engine in the context of tendermint is not

Reviewer: ```suggestion
the benefits of a specific database engine in the context of Tendermint is not
```

particularly well understood, and the maintenance burden for multiple backends

> is not commensurate with the benefit provided. Additionally, because the data

Reviewer: I don't entirely get the connection between the interface being very generic and the in-memory test framework. Is the case that "because we have to be able to implement it in-memory, we can't rely on clever storage behaviour because then we'd have to replicate it?" (If so, I agree)

Author: In the current implementation, yes, the fact that our tests are all mocked out mean that we can't actually rely on anything. In essence, our current storage engine is a `map[[]byte][]bytes{}` that happens to get written to disk sometimes, and all other storage features have to be implemented in the application.

storage layer is handled generically, and most tests run with an in-memory
framework, it's difficult to take advantage of any higher-level features of a
database engine.

Ideally, developers within Tendermint will be able to interact with persisted

> data via a single interface that can function, approximately like an object

Reviewer: I'm not sure a _single_ interface is essential—if we have different data storage requirements (e.g., for logs vs. chain state vs. other things) I think it would be fine to define different interfaces for those. 

I don't know whether we actually have multiple storage concerns, but I do think it's essential that whatever interface(s) we _do_ define reflect the semantics of the consumer, not the provider. That is: The interface should express the domain requirements, not the implementation constraints.

(To be clear: I think this is more or less what you're advocating, not a contrast. And, of course, we have to choose requirements that _can_ be satisfied, but that's achievable)

Author: yeah, I think we're in agreement here. I deleted the word single.

store, and this storage interface will be able to accommodate all existing
persistence workloads (e.g. block storage, local peer management information

> like the "address book", crash-recovery log like the WAL.) In addition to

Reviewer 2: I think moving the WAL to use a shared DB would be a great way to get us out of maintaining a homegrown but critical system. I do wonder, however, what amount of updating to the WAL we want to do within this single project? We could definitely reevaluate the WAL and attempt to remove WAL access where it is not necessary for crash-recovery and correctness, but I worry that a large WAL audit will blow up the scope of the DB change out project.  I think it might be more wise for us to use the new DB as a drop-in replacement for the current WAL in this project.

Reviewer: This is a good question, but if we _do_ use Badger (or similar) it may obviate the need for a WAL: It provides commit semantics that don't require us to maintain our own persistence log. 

Having said that, I don't know much about how our existing WAL is used, so I think it's still worth addressing.

Author: One of the incoming assumptions about this project is that badger's (or similar) would have crash resistance that would obviate the need for the independent WAL. 

The WAL is pretty limited in scope right now, and I would end up doing the audit you propose as part of a more indepth rationalization/modernization of the consensus reactor (which may or may not be obvious.)

providing a more ergonomic interface and new semantics, by selecting a single
storage engine tendermint can use native durability and atomicity features of
the storage engine and simplify its own implementations. 

Data Access Patterns
~~~~~~~~~~~~~~~~~~~~

> Tendermint's data access patterns have the following characteristics:

Reviewer: It seems like writes are more frequent, but maybe smaller (a transaction, an event, etc.) whereas writes are less frequent but larger (e.g., scanning blocks for synchronization, reading event logs). I likewise don't have data to support this impression, but it seems to follow from the way consensus works.

> Tendermint's data access patterns have the following characteristics:

Author: It's also the case that the writes are clustered on "insert" rather than "update" (modulo indexing and some state information) and also the insert pattern (depending on how its keyed/tracked) could be entirely "append-type" writes rather than to random pages. (MVCC systems, like badger, means that writes tend to look append-ish to the disk regardless, perhaps, it's also the case that while we're loggically just appending to the right side of a b-tree (as it were) the indexes make this less true in pract.ce)

> - data size often exceeds memory.

Reviewer: I think you mean aggregate data size here, not the size of individual stored values?

Author: yes, that's correct

- data is rarely mutated after it's written for most data (e.g. blocks), but

>   small amounts of workding data is persisted by nodes and is frequently

Reviewer: ```suggestion
  small amounts of working data is persisted by nodes and is frequently
```

  mutated (e.g. peer information, validator information.)

> - read patterns can be quite random.

Reviewer: Hmm. I don't disagree, but so far it seems like the read patterns are mostly fairly sequential: One might occasionally query for a single transaction (e.g., for debugging) but most of the Tendermint plumbing seems to orient around synchronizing logs between nodes. So the starting-point might be random(ish), but I think after that it's more or less a scan forward in time, right?

Author: I think logically you're correct, but I don't believe that the the pages are likely to be contiguous because the keys/leading bytes aren't monotonically increasing in the way that we'd want. 

I also suspect that for the purposes of consensus all of the data is (basically) is still in the application and we don't hit the db layer, and all other workloads (statesync, blocksync, query load from the application or RPC layers,) are really random.

Reviewer: Yes, perhaps not strictly monotonic—that's a strong claim—but I meant in a sense that we could notionally expect the "next" datum in a sync process to have been recorded "near in time" to the previous one, since transactions are roughly ordered by when they were proposed. So we do have random access, but I think most accesses are (notionally) "within a cache line".

That said, I don't really understand how the app's internal state and the mempool and the persistent log all interact with each other, so I don't have a real difference to propose here.

- crash resistance and crash recovery, provided by write-ahead-logs (in
  consensus, and potentially for the mempool) should allow the system to
  resume work after an unexpected shut down.

Project Goals
~~~~~~~~~~~~~

As we think about replacing the current persistence layer, we should consider
the following high level goals: 

> - drop dependencies on storage engines that have a CGo dependency.

Reviewer: This seems reasonable, but I'm not sure we need to stick too hard on it—if the issue is complicating the build, we can fairly easily set up storage as a lamprey or host so that the implementation can be any old thing.

Author: For sure, I guess in abstract terms my goals center on:

- reduce the build complexity to simplify our testing and development story.
- maintain or reduce the operational complexity for downstream users of tendermint. 
- remove unnecessary genercisism and abstractions 
- push more storage responsibility onto the storage engine and out of tendermint (e.g. WAL/etc.) 

Anything is possible, of course, and it's theoretically all an implementation detail, but I think having these as a framework will lead us to improving the project in useful ways.

- encapsulate data format and data storage from higher-level services
  (e.g. reactors) within tendermint.

> - select a storage engine that does not incur any additional operational

Reviewer: I like this idea, but it seems in tension with the previous desideratum: If we want storage to be encapsulated from the rest of the application, whether it's embedded or not becomes an implementation detail, no?

Author: I agree that its an implementation detail, but I think that it's a good goal to avoid increasing the operational complexity of the system.

Reviewer: Well, I guess that's my point: Embedding a database is potentially quite a big dependency. It'll depend on what we actually decide to use, probably, but it could actually reduce net complexity if we allowed the database to live in an adjacent process with a pipe in between. (I realize "separate process" makes it sound like a hard ops problem, but if we assume adjacency most of the hard parts disappear).

Anyway, not advocating we _should_ do that, only that we should be open to the option.

Author: Today it's separate processes and a pipe, and tomorrow we have a dependency on k8s (this is a joke, mostly.) 

I think I'm open to this kind of solution, I just don't have a good idea/example of what such a system would look like, or what off-the-shelf database might exist that we could use this way.

Reviewer: Well, of course we'd need to be intentional about it. 🙂 

One of the issues I've had on other projects is that storage engines often have complex dependencies, and can pull in half the world to your module file. What has worked well for me is to build a simple pipe-based communication with a separate binary that has all those dependencies, but only needs to be updated for changes to the storage engine rather than triggering dependabot every time some version bump changes one of the many minor dependencies. No need for network routing—you can run either of them as a subprocess of the other, or run them side-by-side with a Unix-domain socket.

Again, though, not advocating for that here, only that in a couple cases I've actually found this _reduced_ the complexity of building, testing, and maintaining my programs. It also means we could use things that require cgo without pulling cgo into the rest of the project. 

If we plan to pick one fairly unstructured storage engine (e.g., Badger) that is already written in Go and doesn't support complex query planning, it may be overkill.

Reviewer 2: Using an embedded db does raise an additional question: Do we want this DB to be accessible to operators of a tendermint node? If we fully embed a DB, all queries on current tendermint state have to be performed through tendermint and all DB tuning is a concern of tendermint itself. I understand the desire to keep operational complexity low, but this does mean a good deal of the observability and tuning is a responsibility of tendermint itself. 

As nodes get larger and larger and more critical, we may worsen operational complexity by not allowing node operators to reach for any of the usual db-related pressure-release valves (sharding, running primary/secondary for improved disaster recovery,  etc.)

Reviewer: That's a good question, but note that both of the database engines we've talked about here (Bolt & Badger) store data on-disk and can be accessed by other tools. (Read-only access while the "owner" process is running, but read-write when it's closed). 

This is still worth considering, but I think it's probably OK if we insist that any write access to the data store be mediated by the consensus node _while the node is running_.

Author: I don't think anything in this project precludes the "seperate storage and core into separate binaries" architecture, which is definitely a useful way of architecting things, but I don't think we're particularly at a point where we'd need that (e.g. the binary size is small, even with badger in it, I don't think the embedded system is particularly onerous or limiting (and is what our users expect, generally.) I operators generally expect the current kind of model, and as a chain developer, I'd be pretty skeptical of a library that required/started its own sidecar binary. Not saying we couldn't do it, but I think CGo and a big binary are easier sells (the no-cgo dependency is mostly for our own build story.)

Additionally, I must confess that I'm not particularly excited (personally) by any of the storage engines that would become available if we relaxed the CGo restriction (e.g. RocksDB, SQLite, etc.)  which might be clouding my opinion above. 

I think these are all good ideas to keep in our pockets.

  complexity (e.g. database should be embedded.)

> - provide databases semantics with sufficient ACID, snapshots, and

Reviewer: ```suggestion
- provide databases semantic with sufficient ACID, snapshots, and
```

I'm not sure what "provide" means here. Are you saying we should let our interfaces rely on those features?

  transactional support.

Open Questions
~~~~~~~~~~~~~~

The following questions remain:

> - what kind of data-access concurrency does tendermint require?

Reviewer: This is a good question. Though if we're assuming Tendermint mediates access to storage through (an) interface(s), this may not matter that much—the node can synchronize access.

If we _don't_ have the node mediate, then the choice of possible backing storage gets much narrower. A lot of popular storage libraries allow concurrent access from a single process, but lock between processes. The ones that don't tend to be more DBMS-like (e.g., Postgres) and not easily embedded. Do you have any particular candidates in mind?

Author: I think Badger is the storage engine to beat for us, though I think we could definitely build what we need on top of bolt (though bolt is less maintained/more simple.) 

Badger's interface/transactions can be used from multiple threads internally so we don't need to add an extra mutex, though you're right, we can if that's needed.

Reviewer: > If blocks are never mutated and are append only do we even need a locking mechanism? We're not going to have a situation of conflicting reads/writes

I presume at some point we will need to prune old data from the log, though. So even if we don't need to worry about concurrent access most of the time, we would presumably need some way to either lock out readers during a pruning stage, or stop-the-world and restart the node?

Reviewer: Re: Badger. Another approach that isn't reflected here is what Redis does. Specifically: Stores and updates on the working set are held in memory, and logged behind to persistent storage (with explicit sync points for operations you need to be sure landed on disk). Possibly we can't do this if our working set is expected to be too big for memory, but the strategy seems like a good possible fit for our usage patterns.

Author: > If blocks are never mutated and are append only do we even need a locking mechanism? We're not going to have a situation of conflicting reads/writes

Blocks aren't modified, indexes and various caches presumably will. (e.g. if we persist pointers on disk to "latest" things, as well as the WAL. 

I guess what I'm arguing for is more that tendermint should rely on the database engine's concurrency control and shouldn't need to add an additional mutex around the database. 

---- 

Redis is delightfully simple, and I think there is merit to their approach, from a high level, but I think a big part of this whole project is to reduce the current level of complexity in the relationship between the consensus engine (tendermint) and the storage engine while also reducing (somewhat) the porous boundary between these components. It feels like doing "the redis thing" would be something of a premature optimization at this point.

- would tendermint users SDK/etc. benefit from some shared database
  infrastructure?
  

>   - In earlier conversations it seemed as if the SDK has selected Badger and

Reviewer: Related to the above: Badger doesn't allow concurrent writes from multiple processes. I am not sure about Rocks.

Author: but they do allow concurrent writes from within a single processes (e.g. multiple threads) without requiring us to put a lock in front of it. 

in this case, the application and tendermint would be in the same process, so I think we're ok.

Reviewer: > in this case, the application and tendermint would be in the same process, so I think we're ok.

The case I was thinking of is indexing: One of the main reasons we were asked to improve the schema for the psql indexing sink was so that external consumers could more easily do queries against the data. That presumes they'd be accessing the data store independently from the node and its application.

If the data store we're talking about here is _not_ for indexing, then this may be moot.

Author: A lot of database systems make default caching and related decisions based on the assumption that there is only one storage engine running on a machine, and if the SDK and tendermint both have independent DBs it's possible that they could clobber each other. 

I'm not sure this would happen, but it's good to have something like this on our radar just in case

    RocksDB for their storage engines, and it might make sense to be able to
    (optionally) pass a handle to a Badger instance between the libraries in
    some cases.

> - what are typical data sizes, and what kinds of memory sizes can we expect

Reviewer: Yes! This is an interesting data question.

  operators to be able to provide?

- in addition to simple persistence, what kind of additional semantics would
  tendermint like to enjoy (e.g. transactional semantics, unique constraints,
  indexes, in-place-updates, etc.)?

Decision Framework
~~~~~~~~~~~~~~~~~~

Given the constraint of removing the CGo dependency, the decision is between
"badger" and "boltdb" (in the form of the etcd/CoreOS fork,) as low level. On
top of this and somewhat orthogonally, we must also decide on the interface to
the database and how the larger application will have to interact with the
database layer. Users of the data layer shouldn't ever need to interact with

> raw byte slices from the database, and should mostly have the experience of

Reviewer: I generally agree with this approach, but I think we need to be careful: This implies anything that ever wants to store or read data needs to depend either on those same Go types, which makes it harder to encapsulate the implementation.

I'd be somewhat more comfortable if we define the storage layer in terms of, say, protobuf wire format encodings. Then the dependency is only on the protobuf schema, and any usage outside the core (e.g., for debugging tools) can generate its own wrapper types confident in the ability to unmarshal the stored data forward-compatibly.

I know this seems like a minor point, and it mostly is, but strictly-typed interfaces have a tendency to pull one into a dependency hairball.

Author: I think we're in agreement. The sentence says "we shouldn't just stuff byte slices in the db" (which is roughly what we currently do,) and I think the question it's raising is the question you're raising (correctly.)

interacting with Go-types.

> Badger is more consistently developed and has a broader featureset than

Reviewer: ```suggestion
Badger is more consistently developed and has a broader feature set than
```

Bolt. At the same time, Badger is likely more memory intensive and may have

> more overhead in terms of open file handles given it's model. At first glance,

Reviewer: I have also empiricially found the performance of Badger to be significantly better than Bolt, on all the workloads I've tried (my use cases aren't the same as Tendermint's, so I doubt my benchmarks are meaningful, but both reads and writes were faster for my random-access workloads).

Author: yeah, this is consistent with my expectation.

Badger is the obvious choice: it's actively developed and it has a lot of
features that could be useful. Bolt is not without some benefits: it's stable
and is maintained by the etcd folks, it's simpler model (single memory mapped
file, etc,) may be easier to reason about.

I propose that we consider the following specific questions about storage
engines:

> - does Badger's evolving development, which may result in data file format

Reviewer: The Badger developers have some [compatibility rules](https://github.com/dgraph-io/badger#choosing-a-version) that they have so far respected pretty well: They won't make breaking data format changes without a major version update. They do reserve the right to make some kinds of breaking Go API changes, but I think those are less of a problem for us.

  changes in the future, and could restrict our access to using the latest

>   version of the library between major upgrades, present a problem?

Author: indeed, but we'll have to manage and handle our relationship with our own upstream library.

> - do we do we have goals/concerns about memory footprint that Badger may

Reviewer: Badger has pretty good knobs for tuning how much of the indexing structure of the underlying sstables is kept in RAM. It's the usual time-memory tradeoff, but I think we shouldn't be at too much risk of blowing out our RAM budget in most realistic environments.

Reviewer 2: I generally think trying to keep the number of user facing knobs low is a good practice for Tendermint but I do worry about us picking the values and OOMing in low memory environments. I think we'd be fine picking some simple specs for validators to target, but it's worth considering.

As nodes grow, so will the size of the index, so it's likely worth taking a little bit of time for us to understand that relationship for the storage engine/indexing scheme that we choose.

Reviewer: I agree with this, re: configuration knobs. That said: We don't have to choose between "no knobs" and "the full generality of the storage engine". We can add a setting that lets the user specify a size hint, and "compile" that into appropriate storage settings ourselves (with a "do something sensible" default). If that turns out to be inadequate, we can always add more knobs later, but this lets us walk that line slowly.

  prevent us from hitting, particularly as data sets grow over time?

> - what kind of additional tooling might we need/like to build (dump/restore,

Reviewer 2: Badger does appear to have a [Backup](https://github.com/dgraph-io/badger/blob/master/backup.go#L49) method that we could leverage, I'm not sure we need much more advanced for our usecase.

  etc.)?

- do we want to run unit/integration tests against a data files on disk rather

>   than relying exclusively on the memory database?

Reviewer 3: If there's a distinction in the characteristic of in-memory vs persistent databases and the integration tests are affected by such differences then I think that makes sense - but I think that's a very small subset of tests

Reviewer: If we are really going to commit to a single implementation target, I think we shouldn't try to do anything fancy here: We should test the production code. One nice property of Badger is that it is almost entirely plain files on disk (it does create some advisory locks, but nothing fancy with device files or sockets as far as I know).

Reviewer 2: I think our integration tests should certainly use the full on-disk storage engine. I would advocate that, where possible, we mock it out in the unit tests for the sake of simplicity and speed. For example, consensus tests that are not testing crash recovery do not require access to a real WAL and can and should use a mock (likely one that's implemented as a no-op).

Reviewer: If we were talking about a heavy-weight database engine like Postgres, I'd agree. But the storage engines we've been considering are lightweight and fast: Fast enough that if we need to mock them out to make the tests go faster, we have probably already failed our performance goals in production.

So I don't think we can or should try to mock them in this case, though I do agree with the impulse more generally for "large" service dependencies.

Reviewer 2: My concern here is mostly around the pain associated with adding and running more tests. If we're confident that the embedded DB can be fast enough that the slowdown of adding a new test is minimal and that we can easily write default setup/teardown code that makes adding DB access into a test very straightforward, then my concerns are largely unfounded.

Reviewer: > My concern here is mostly around the pain associated with adding and running more tests. 

That's quite valid, but for these examples it wouldn't be any harder to add additional tests using the real production code than it would using a mock. 

I think it's best to avoid test-doubles unless using the production code is both (a) too complicated to do in a test environment (e.g., requires credentials, needs to spin up a separate service, etc.), and (b) not relevant enough to the code under test to be worth doing.

Here, I don't think we have either issue: Setting up a database is a couple lines of code, and the correct behaviour of the data store is relevant to anything tested against it. With the options we've been discussing, we could even just check in small databases literally as testdata, they're small and self-contained; creating new ones is just a few lines of code. 

Personally, I'd rather provide a support library (a la `httptest`) than have to deal with yet another mock that I can't be sure is actually implementing the correct semantics.

Author: Woah, this thread got out of control, and I think this was kind of a throw away point from my perspective. 

I mostly wanted to point out that except for the e2e tests, the database code is never run or used in our testing environment, and that's a flaw for a collection of reasons:
- because we should be able to run tests with more realistic situations when it make sense
- the current tests, because they're written against a hashmap, may miss some key behavior

This project will definitely entail building a new set of testing tools that will have to be fast/etc.

Project Scope
~~~~~~~~~~~~~

This project will consist of the following aspects:

- selecting a storage engine, and modifying the tendermint codebase to

>   disallow any configuration of the storage engine outside of the tendermint. 

Reviewer 3: One of the modifications I would like to see is to merge the state and block store and make the committing of a new block atomic

Author: This seems like a good goal, and one that I think fits within existing aspects of this project as described ("have a database with sufficent atomicity semantics" and "update core tendermint code to use the new database"). 

I'm not sure how to add it to this document in a way that isn't a list of things to refactor (e.g. I've not put "delete the WAL implementation" as a goal/scope item, though clearly that's the case here.)
